server/src/main/java/org/elasticsearch/action/support/tasks/TransportTasksAction.java

```diff
@@ -85,66 +81,164 @@ protected TransportTasksAction(
-        TasksRequest request = nodeTaskRequest.tasksRequest;
-        processTasks(request, ActionListener.wrap(tasks -> nodeOperation(task, listener, request, tasks), listener::onFailure));
+        final var taskResponses = new ArrayList<TaskResponse>();
+        final var taskOperationFailures = new ArrayList<TaskOperationFailure>();
+        final var failedNodeExceptions = new ArrayList<FailedNodeException>();
+
+        final var resultListener = new SubscribableListener<TasksResponse>();
+        final var resultListenerCompleter = new RunOnce(() -> {
+            if (task instanceof CancellableTask cancellableTask && cancellableTask.notifyIfCancelled(resultListener)) {
+                return;
+            }
+            // ref releases all happen-before here so no need to be synchronized
+            ActionListener.completeWith(
+                resultListener,
+                () -> newResponse(request, taskResponses, taskOperationFailures, failedNodeExceptions)
+            );
+        });
+
+        // collects node listeners & completes them if cancelled
+        final var nodeCancellationListener = new SubscribableListener<NodeTasksResponse>();
+        if (task instanceof CancellableTask cancellableTask) {
+            cancellableTask.addListener(() -> {
+                assert cancellableTask.isCancelled();
+                resultListenerCompleter.run();
+                cancellableTask.notifyIfCancelled(nodeCancellationListener);
+            });
+        }
+
+        try (var refs = new RefCountingRunnable(() -> {
+            resultListener.addListener(listener);
+            resultListenerCompleter.run();
+        })) {
+            final var discoveryNodes = clusterService.state().nodes();
+            final String[] nodeIds = resolveNodes(request, discoveryNodes);
+
+            final var transportRequestOptions = TransportRequestOptions.timeout(request.getTimeout());
+
+            for (final var nodeId : nodeIds) {
+                final ActionListener<NodeTasksResponse> nodeResponseListener = ActionListener.notifyOnce(new ActionListener<>() {
+                    @Override
+                    public void onResponse(NodeTasksResponse nodeResponse) {
+                        synchronized (taskResponses) {
+                            taskResponses.addAll(nodeResponse.results);
+                        }
+                        synchronized (taskOperationFailures) {
+                            taskOperationFailures.addAll(nodeResponse.exceptions);
+                        }
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        if (task instanceof CancellableTask cancellableTask && cancellableTask.isCancelled()) {
+                            return;
+                        }
+
+                        logger.debug(Strings.format("failed to execute on node [{}]", nodeId), e);
```

Comment:
I was wondering if can perform formatting lazily by using a supplier : `logger.debug(() -> Strings.format(`

Reply:
Yes; it's not very common to get here, but I'll do that.